reflect.go

```diff
@@ -0,0 +1,387 @@
+package main
+
+import (
+	"fmt"
+	"go/types"
+	"path/filepath"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+	"golang.org/x/tools/go/ssa"
+)
+
+var checkedAPIs = make(map[string]bool)
+
+// Record all instances of reflection use, and don't obfuscate types which are used in reflection.
+func (tf *transformer) recordReflection(ssaPkg *ssa.Package) {
+	if reflectSkipPkg[ssaPkg.Pkg.Name()] {
+		return
+	}
+
+	lenPrevKnownReflectAPIs := len(cachedOutput.KnownReflectAPIs)
+
+	// find all unchecked APIs to add them to checkedAPIs after the pass
+	notCheckedAPIs := make(map[string]bool)
+	for _, knownAPI := range maps.Keys(cachedOutput.KnownReflectAPIs) {
+		if !checkedAPIs[knownAPI] {
+			notCheckedAPIs[knownAPI] = true
+		}
+	}
+
+	tf.ignoreReflectedTypes(ssaPkg)
+
+	// all previously unchecked APIs have now been checked add them to checkedAPIs,
+	// to avoid checking them twice
+	maps.Copy(checkedAPIs, notCheckedAPIs)
+
+	// if a new reflectAPI is found we need to Re-evaluate all functions which might be using that API
+	if len(cachedOutput.KnownReflectAPIs) > lenPrevKnownReflectAPIs {
+		tf.recordReflection(ssaPkg)
+	}
+}
+
+func (tf *transformer) ignoreReflectedTypes(ssaPkg *ssa.Package) {
+	for _, memb := range ssaPkg.Members {
+		if t, ok := memb.(*ssa.Type); ok {
+			// methods aren't package members only their reciever types are
+			// so some logic is required to find the methods a type has
+
+			// yes, finding all methods really only works with both loops
+			mset := ssaPkg.Prog.MethodSets.MethodSet(t.Type())
+			for i, n := 0, mset.Len(); i < n; i++ {
+				fun := ssaPkg.Prog.MethodValue(mset.At(i))
+				if fun != nil {
+					tf.checkCalls(fun)
+				}
+			}
+
+			mset = ssaPkg.Prog.MethodSets.MethodSet(types.NewPointer(t.Type()))
+			for i, n := 0, mset.Len(); i < n; i++ {
+				fun := ssaPkg.Prog.MethodValue(mset.At(i))
+				if fun != nil {
+					tf.checkCalls(fun)
+				}
+			}
+		}
+
+		if fun, ok := memb.(*ssa.Function); ok {
+			// these not only include top level functions, but also synthetic
+			// functions like the initialization of global variables
+			tf.checkCalls(fun)
+		}
+	}
+}
+
+func (tf *transformer) checkCalls(fun *ssa.Function) {
+	if fun.Synthetic == "loaded from gc object file" {
+		// fun.WriteTo crashes otherwise
+		return
+	}
```

Comment:
this ought to be a TODO; either we need better logic in our code, or we should raise a bug upstream to avoid that crash.

Reply:
This is still a problem needs to be fixed upstream